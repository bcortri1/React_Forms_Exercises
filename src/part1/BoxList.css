.BoxList {
    display: flex;
    background-color: white;
    justify-content: center;
    flex-wrap: wrap;
    margin: 20px;
}