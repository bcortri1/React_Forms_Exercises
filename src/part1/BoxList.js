import React, { useState } from 'react';
import './BoxList.css'
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import { v4 as uuid } from 'uuid';


const BoxList = () => {
    const initialState = []
    const [boxArr, setBoxList] = useState(initialState);

    const removeBox = (id) => {
        setBoxList(boxArr.filter((box) => (box.id !== id)));
    }

    const addBox = (data) => {
        let id = uuid();
        setBoxList([...boxArr, {...data, id}]);
    }

    return <div>
        <div className='BoxList'>
            <NewBoxForm addBox={addBox}/>
        </div>
        <div className='BoxList'>
            {boxArr.map((props) =>
                <Box key={props.id} {...props} removeBox={removeBox} />
            )}
        </div>
    </div>;
};

export default BoxList;