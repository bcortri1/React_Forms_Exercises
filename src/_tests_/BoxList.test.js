import React from "react";
import { render, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom';
import BoxList from "../part1/BoxList";



// smoke test
it("renders without crashing", function () {
    render(<BoxList />);
});

// snapshot test
it("matches snapshot", function () {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("add new Boxes", function () {
    const { queryByText } = render(<BoxList />);
    const addButton = queryByText("Add!");
    fireEvent.click(addButton);

    expect(document.getElementsByClassName("Box")[0]).toBeInTheDocument();
    expect(document.getElementsByClassName("Box")[1]).not.toBeDefined();

});

it("can add and contain multiple Boxes", function () {
    const { queryByText, getByLabelText } = render(<BoxList />)
    const width = getByLabelText("Width:");
    const height = getByLabelText("Height:");
    const color = getByLabelText("Color:");
    const addButton = queryByText("Add!");

    fireEvent.click(addButton);
    expect(document.getElementsByClassName("Box")[0]).toBeInTheDocument();
    expect(document.getElementsByClassName("Box")[0].style.width).toEqual("50px");
    expect(document.getElementsByClassName("Box")[0].style.height).toEqual("50px");
    expect(document.getElementsByClassName("Box")[0].style["background-color"]).toEqual("rgb(0, 0, 0)");

    fireEvent.change(width, { target: { value: "100" } });
    fireEvent.change(height, { target: { value: "900" } });
    fireEvent.change(color, { target: { value: "#911111" } });
    fireEvent.click(addButton);
    expect(document.getElementsByClassName("Box")[1]).toBeInTheDocument();
    expect(document.getElementsByClassName("Box")[1].style.width).toEqual("100px");
    expect(document.getElementsByClassName("Box")[1].style.height).toEqual("900px");
    expect(document.getElementsByClassName("Box")[1].style["background-color"]).toEqual("rgb(145, 17, 17)");
    expect(document.getElementsByClassName("Box")[2]).not.toBeDefined();

});
