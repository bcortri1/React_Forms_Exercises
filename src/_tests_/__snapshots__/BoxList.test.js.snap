// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="BoxList"
    >
      <form
        id="new-box-form"
      >
        <label
          for="width"
        >
          Width:
        </label>
        <input
          id="width"
          name="width"
          required=""
          type="number"
          value="50"
        />
        <label
          for="height"
        >
          Height:
        </label>
        <input
          id="height"
          name="height"
          required=""
          type="number"
          value="50"
        />
        <label
          for="color"
        >
          Color:
        </label>
        <input
          id="color"
          name="color"
          type="color"
          value="#000000"
        />
        <button
          id="form-btn"
        >
          Add!
        </button>
      </form>
    </div>
    <div
      class="BoxList"
    />
  </div>
</DocumentFragment>
`;
